pkg/provider/azure_controller_common.go

```diff
@@ -267,7 +281,46 @@ func (c *controllerCommon) AttachDisk(ctx context.Context, async bool, diskName,
 			klog.Warningf("azureDisk - switch to batch operation due to rate limited, QPS: %f", c.diskOpRateLimiter.QPS())
 		}
 	}
-	return lun, vmset.WaitForUpdateResult(ctx, future, nodeName, "attach_disk")
+
+	err = c.waitForUpdateResult(ctx, vmset, nodeName, future, err)
+
+	// if no error was returned, attach was successful
+	if err == nil {
+		klog.V(2).Infof("azuredisk - successfully attached disks to node %s: %s", nodeName, diskMap)
+	} else {
+		return -1, err
+	}
+
+	return lun, nil
+}
+
+// waitForUpdateResult handles asynchronous VM update operations and retries with backoff if OperationPreempted error is observed
+func (c *controllerCommon) waitForUpdateResult(ctx context.Context, vmset VMSet, nodeName types.NodeName, future *azure.Future, updateErr error) (err error) {
+	err = updateErr
+
+	if err == nil {
+		err = vmset.WaitForUpdateResult(ctx, future, nodeName, "attach_disk")
+	}
+
+	if vmUpdateRequired(future, err) {
+		if derr := kwait.ExponentialBackoffWithContext(ctx, attachBackoff, func() (bool, error) {
```

Comment:
why not use `err = kwait.ExponentialBackoffWithContext` directly here, then you don't need to `err = derr` in L314

Reply:
That's because we can't return `err` directly in the condition func of `kwait.ExponentialBackoffWithContext` as returning a non-nil error would break the loop.